llmfoundry/callbacks/curriculum_learning_callback.py

```diff
@@ -7,47 +7,225 @@
+                'max_duration.'
+            ))
+
+        self._validate_dataloader(state.train_dataloader)
+
+    def after_load(self, state: State, logger: Logger):
+        del logger
+
+        self._validate_dataloader(state.train_dataloader)
+
+        # If checkpoint was saved before iteration was incremented, we need to increment it now
+        if ((
+            self._schedule[self._schedule_index]['duration'].unit
+            == TimeUnit.TOKEN and state.timestamp.token_in_iteration
+            >= self._schedule[self._schedule_index]['duration'].value
+        ) or (
+            self._schedule[self._schedule_index]['duration'].unit
+            == TimeUnit.EPOCH and state.timestamp.epoch_in_iteration
+            >= self._schedule[self._schedule_index]['duration'].value
+        )):
+            log.error((
+                'The CurriculumLearning callback has detected that the previous run did not correctly '
+                'increment the iteration.'
+            ))
+            self._schedule_index += 1
+            state.timestamp = state.timestamp.to_next_iteration()
+
+    def iteration_start(self, state: State, logger: Logger):
+
+        # Swap the dataset if starting a new iteration that's not the original datamix
+        if self._schedule_index > 0:
+            # Reset and initialize state train dataloader
+            log.info(
+                'trainer._train_data_spec should be updated whenever the dataloader is updated',
+            )
+            # Dataloaders with the same prefix access the same shared memory
+            # which is stale
+            clean_stale_shared_memory()
+            datamix = copy.deepcopy(self._schedule[self._schedule_index])
+            data_spec = self._build_train_loader(
+                train_loader_config=datamix['train_loader'],
+                logger=logger,
+            )
+            state.set_dataloader(
+                dataloader=data_spec.dataloader,
+                dataloader_label='train',
+            )
+            state.train_dataloader = state.dataloader
+            self._validate_dataloader(state.train_dataloader)
+
+        # Set the length of the new iteration
+        state._iteration_length = self._schedule[self._schedule_index
+                                                ]['duration']
+
+    def iteration_end(self, state: State, logger: Logger):
+        del state, logger  # unused
+
+        self._schedule_index += 1
+
+    def state_dict(self):
+        return {
+            'schedule': self._schedule,
+            'schedule_index': self._schedule_index,
+        }
+
+    def load_state_dict(self, state: dict[str, Any]):
+        self._schedule_index = state['schedule_index']
+
+        # Ensure that the schedule has not changed on previously trained datamixes
+        for idx in range(state['schedule_index']):
+            assert self._schedule[idx] == state['schedule'][idx]
```

Comment:
can we instead do if + runtime error giving a helpful msg to user?

Reply:
I don't think this is critical to application logic since behavior will be exactly the same with and without assert
https://google.github.io/styleguide/pyguide.html#244-decision

Reply:
this should be an error not an assert (https://github.com/mosaicml/composer/blob/dev/STYLE_GUIDE.md#4-use-of-assert). behavior is not unchanged without the assert, it would silently do something bad (or crash somewhere else, not certain)

Reply:
as with the other places

Reply:
The callback doesn't use previous indices in the schedule so it wouldn't change its behavior. If the user changes the previous datamixes, the callback resumes on the current datamix from the checkpoint and ignores everything previously in the schedule.

Reply:
Done